pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java

```diff
@@ -88,8 +95,19 @@ public void init(ExternalView externalView, IdealState idealState, Set<String> o
         _partitionInfoMap.put(segment, partitionInfo);
       }
     }
+    rebuildReverseMap();
   }
 
+  private void rebuildReverseMap() {
+    _partitionInfoSegments.clear();
+    Set<String> segmentsFound = new HashSet<>();
+    _partitionInfoMap.forEach((segmentName, partitionInfo) -> {
+      segmentsFound.add(segmentName);
+      Set<String> segments = _partitionInfoSegments.computeIfAbsent(partitionInfo, (info) -> new HashSet<>());
```

Comment:
This won't be correct if a segment contains more than 1 partitions. We need to maintain a `Map<PartitionFunction, Map<Integer, Set<String>>>`, where the inner map can be optimized to a `List<Set<String>>`, and the index of the list is the partition number.